### ISSUE TITLE
Better handling at the connection drop?

### ISSUE THREAD

**Issue by highfield** [nats-io/nats.net.v1#184]
Since I'm wondering how the client/server behave on a remote (not-so-realiable) connection, I started some test.
My current attempt is a single .Net Core 2.0 console app, which hosts both "master" and "slave" endpoints: both use this driver (latest stable release from NuGet), and connect against a single server on the same machine (Win10). The aim is perform a kind of query (master queries slave) using Nats.io as broker.
When the app starts, both the master and the slave connect to the broker, which run in verbose mode thus I can see what happens. The problem appears when I close the Nats.io server: immediately, the CPU usage jumps from 0 to 25%. If I restart the broker nothing happens until 120 secs elapse (I believe that's the ping/pong period).
From what I understand, there's a huge impact on the "Read" method, within the "readLoop" function, "Conn.cs" module. It seems that there's no way to cancel the block using a cancelable ReadAsync. By closing the NetworkStream upon a connection drop, probably would speed up the detection.
Is there any better solution to handle a connection drop, especially in order to avoid an excessive CPU usage?
...diggin' into in the meantime...

**Comment by ColinSullivan1** [comment]
@highfield , thank you for raising this.  How are you closing the NATS server?  This sounds like the underlying TCP connection has dropped and the client or OS has not detected the connection has gone at the TCP level.  Have you added a [DisconnectedEventHandler](http://nats-io.github.io/csharp-nats/class_n_a_t_s_1_1_client_1_1_options.html#a6116c0a92a645bbc356bde9533220ea4) to see if the client is detecting a disconnection immediately? More on handlers can be found in this [section](https://github.com/nats-io/csharp-nats#advanced-usage) of the readme.

**Comment by highfield** [comment]
@ColinSullivan1 the server is started by a PowerShell console, then halted via ctrl-C: very easy to start/stop this way.
I hadn't time enough to make more experiments, other than taking the ping interval much shorter than the default. Once I set to 5 secs, the CPU load is very short (shorter than the interval), and the system responsive. Of course, that imply a small overhead on the network.
I definitely will try your suggestion about the event: honestly, I didn't realize that but sounds promising. I can't do anything right now, because I'm at home.
Many thanks in the meantime!

**Comment by ColinSullivan1** [comment]
@highfield , any news on this issue?

**Comment by highfield** [comment]
@ColinSullivan1 shame on me for the delay!
Well, I basically forked the project and have rewritten most from scratch. I only hope you won't hurt because that! The original base is and will be by Apcera, as well copyright and license.
My work is almost done, and it will be online (public) in a few weeks. It had took much more time than expected because I must manage other projects at same time.

**Comment by ColinSullivan1** [comment]
@highfield , No hard feelings at all.  Just following up - is this your assembly [here](https://github.com/highfield/Cet.NATS.Client)?  Looks great, and we can certainly link to it on our website.

**Comment by Sergey-Terekhin** [comment]
Hi!
We've encountered the same issue - do you have any plans to merge @highfield  solution into your client? I'd prefer to use official client rather than not

**Comment by ghstahl** [comment]
I don't know if this is related, but I am using the pub/sub examples where I modified it to take in a queue group.  Everything works, so let's do some experiments.   My NATS server is running on windows 10 kubernetes via docker desktop.  I did a port-foward so that my subscriber and publisher can connect.  After sending a few messages through I kill the port-foward.  The subscriber never reconnections after I bring back the port-forward.  Restarting the subscriber manually hooks it back up again.  This is all useing dotnetcore 3.0.
